File: Easy/longest_common_prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
  if (strs.length == 0 || strs.length > 200) {
    return "";
  }

  strs = strs.sort((a, b) => a.length - b.length)
  const lenShortText = strs[0].length;
  let collectPrefix = [];

  for (let i = 0; i < lenShortText; i++) {
    if(collectPrefix.length == i) {
        collectPrefix.push(strs[0][i]);
    }

    if (i > collectPrefix.length || collectPrefix.length == 0) {
        break;
    }

    for (let j = 0; j < strs.length; j++) {
        if (strs[j] === "") {
            continue;
        }

         if (strs[j][i] !== collectPrefix[i]) {
            collectPrefix.pop();
            break;
        }
    }
  }

  return collectPrefix.join("");
};

console.log(longestCommonPrefix(["flower","flow","flight"])); //fl
console.log(longestCommonPrefix(["flower","flow",""])); //fl
console.log(longestCommonPrefix(["flower","flower","flower","flower"])); //flower
console.log(longestCommonPrefix(["ab", "a"])); //a
console.log(longestCommonPrefix(["", ""])); //
console.log(longestCommonPrefix(["dog","racecar","car"])); //
console.log(longestCommonPrefix(["a"])); //a

File: Easy/longest_common_prefix.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 || len(strs) > 200 {
		return ""
	}

	sort.Slice(strs, func(i, j int) bool {
		return len(strs[i]) < len(strs[j])
	})

	lenShortText := len(strs[0])
	collectPrefix := []string{}

	for i := 0; i < lenShortText; i++ {
		if len(collectPrefix) == i {
			collectPrefix = append(collectPrefix, string(strs[0][i]))
		}

		if i > len(collectPrefix) || len(collectPrefix) == 0 {
			break
		}

		for _, text := range strs {
			if text == "" {
				continue
			}

			if string(text[i]) != collectPrefix[i] {
				collectPrefix = collectPrefix[:len(collectPrefix)-1]
				break
			}
		}
	}

	return strings.Join(collectPrefix, "")
}

func run_longestCommonPrefix() {
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"}))             // fl
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", ""}))                   // fl
	fmt.Println(longestCommonPrefix([]string{"flower", "flower", "flower", "flower"})) // flower
	fmt.Println(longestCommonPrefix([]string{"ab", "a"}))                              // a
	fmt.Println(longestCommonPrefix([]string{"", ""}))                                 //
	fmt.Println(longestCommonPrefix([]string{"dog", "racecar", "car"}))                //
	fmt.Println(longestCommonPrefix([]string{"a"}))                                    // a
}
